Proximity component

nickwaring opened the pull request:
Component to monitor the proximity of devices to a particular zone. The
result is an entity which maintains the proximity data.
Configuration.yaml is used to provide settings such as zone to monitor
and devices to track.

nickwaring committed:
device presence

Component to monitor the proximity of devices to a particular zone. The
result is an entity which maintains the proximity data.
Configuration.yaml is used to provide settings such as zone to monitor
and devices to track.

balloob commented:
You should move it from the `device_tracker` folder to the root of the components folder, otherwise it will be unable to run.

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+
+        """========================================================"""
+        #if the device is not the closest to the proximity zone
+        if device_is_closest_to_zone == False:
```
```python
if not device_is_closest_to_zone:
```

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+        """========================================================"""
+        #calculate direction of travel
+        #stop if we cannot calculate the direction of travel (i.e. we don't have a previous state and a current LAT and LONG)
+        if old_state == None or new_state.attributes['latitude'] == None or not('latitude' in old_state.attributes):
```
```python
if old_state is None or …
```

The test `new_state.attributes['latitude'] == None` will fail, the safe guard is the check after it. You can basically remove this specific test.

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+        #calculate direction of travel
+        #stop if we cannot calculate the direction of travel (i.e. we don't have a previous state and a current LAT and LONG)
+        if old_state == None or new_state.attributes['latitude'] == None or not('latitude' in old_state.attributes):
+            entity_attributes = {ATTR_DIST_FROM:distance_from_zone, ATTR_DIR_OF_TRAVEL:'Unknown', ATTR_NEAREST_DEVICE:entity_name, ATTR_HIDDEN: False} 
```
So there is a 79 character limit per line :heart:

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+        """========================================================"""
+        #update the proximity entity
+        entity_attributes = {ATTR_DIST_FROM:round(distance_from_zone), ATTR_DIR_OF_TRAVEL:direction_of_travel, ATTR_NEAREST_DEVICE:entity_name, ATTR_HIDDEN: False} 
+        hass.states.set(ENTITY_ID, round(distance_from_zone), entity_attributes)
```
You shouldn't use `hass.states.set`. That will be used internally by your Proximity Entity. You would want to update the attributes of your entity and call `proximity.update_ha_state()`:

```python
proximity._dist_from = distance_from_zone
…
proximity.update_ha_state()
```

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+
+Use configuration.yaml to enable the user to easily tune a number of settings:
+- Zone: the zone to which this component is measuring the distance to. Default is the home zone
+- Override Zones: where proximity is not calculated (e.g. work or school)
```
Shouldn't this be called ignore zones ? Because they don't override anything in this component.

balloob commented on homeassistant/components/device_tracker/proximity.py:
```diff
@@ -0,0 +1,279 @@
+
+        """========================================================"""
+        #Debug lines to aid testing & fault diagnosis
+        if not(old_state == None):
```
```python
if old_state is not None:
```

balloob commented:
I am still not quite clear what exactly the use cases are for this component. Could you elaborate on the different states the proximity entity can be?

balloob commented:
I've given some Python hints as I know that you're new to Python.

balloob commented:
Also, CI found a few [style violations](https://travis-ci.org/balloob/home-assistant/jobs/105537650#L203-L395).

Bart274 commented:
Hi balloob,

I've been talking with @nickwaring about this component and I think it can be very useful.
With this component you can monitor who's closest to a certain zone and if that person if driving towards that zone or getting away from it. you can use this for example to adjust your thermostat's settings based on whether or not someone is getting closer to home so it will be nice and warm when that person arrives at home. if you want to do this without this component, you'll need to create multiple concentric zones in order to adjust the temperature accordingly.
Another example could be that if your alarm sounds, you can notify the person closest to home because he/she can go and check things out the quickest.
I was thinking of a second version of this component where you compare the distance of one device to several zones so you'll see what the closest zone is to that device. I have created 12 zones now. When I'm driving around with my baby and something happens (like he throws up or something) I would love to open up HA real quick and see if I'm closest to my home or to my parents, to my brothers, grandparents, the parents of my girlfriend,... so I can see easily to where I could drive to get help the quickest way or who I could call in case of an emergency.

Bart274 committed:
Delete proximity.py

Bart274 committed:
Create proximity.py

nickwaring committed:
Merge pull request #1 from Bart274/patch-2

Create proximity.py

nickwaring committed:
Merge pull request #2 from Bart274/patch-1

Delete proximity.py

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,337 @@
+				device_state.attributes['longitude'])
+
+            # compare the distances from the proximity zone
+            if distance_from_zone < compare_distance_from_zone:
```
These 3 if statements hold no purpose. You can rewrite it as this and save like 15 lines of code.
```python
if distance_from_zone >= compare_distance_from_zone:
    device_is_closest_to_zone = False
    break
```

nickwaring replied:
amazing!

nickwaring commented:
Hi, @balloob 

Regarding the use case... we exchanged posts a while ago and it was from your suggestion that I embarked on the task of creating the component (I don't expect you to remember BTW).

My use of HA started with a series of zones with increasing radiuses to allow me to regulate my nest thermostat as I travelled away from, or closer to, home. Initially this worked well but soon the ability to know the direction of travel and make adjustments required ever increasing numbers of rules (multiple devices, multiple zones, not_home only triggering when outside all zones and handling when a device "skipped" a zone due to a reporting drop-out). Therefore, this component provides a simple state and attributes which can be used to support any proximity based automation task.

Many many thanks for the advice and guidance. It took me a while to get my head around removing the nested if.. else... statements but once I'd made the mental leap I can see the sense (aka you were right)!

@Bart274 has been an enormous help in getting this component to where it is - thanks bart!.

nickwaring committed:
Update proximity.py

nickwaring committed:
Update proximity.py

nickwaring committed:
Update proximity.py

Minor amendment to else statement

nickwaring committed:
fixed else statement

nickwaring committed:
amended entity update approach to use update_ha_state()

nickwaring committed:
updated formatting errors reported by travis

nickwaring committed:
formatting updates to confirm to code standards

nickwaring committed:
formatting changes to remove pylint errors

Bart274 committed:
changes to remove some travis build errors

Bart274 committed:
some more changes to remove travis build errors

Bart274 committed:
still some errors

Bart274 committed:
next try

Bart274 committed:
missing a space for the inline comment

nickwaring committed:
removed nearest_device attribute

nickwaring committed:
removed new line at end of file

nickwaring committed:
added final line at EOF

nickwaring committed:
updates to comments

nickwaring committed:
fix travis errors

Bart274 committed:
Merge remote-tracking branch 'refs/remotes/balloob/dev' into dev

Bart274 committed:
fixed some typos

Bart274 committed:
Component was always Hidden

Bart274 committed:
remove Hidden

Now Hidden was always set to False so you couldn't change it in
customize, now you can

Bart274 committed:
Merge remote-tracking branch 'refs/remotes/balloob/dev' into dev

Bart274 committed:
Re-added nearest device

Now the nearest device will display all the devices that are at home

Bart274 committed:
fixes travis build errors

Bart274 committed:
too long line split up

Bart274 committed:
wrong syntax error fixed

Bart274 committed:
Merge remote-tracking branch 'refs/remotes/balloob/dev' into dev

Bart274 committed:
Add first version of test file of proximity component

Bart274 committed:
change test + change logic in update

Now it will reset the distance etc to 'not set' when all devices are in
ignored zones.
Each time a device changes state it will recalculate which device is now
closest to the monitored zone. When a new device is closest to the zone
and it isn't the one being updated the direction of travel is unknown.

Bart274 committed:
fix Travis CI Build errors

Bart274 committed:
Merge remote-tracking branch 'refs/remotes/balloob/dev' into dev

Bart274 committed:
wrong indentation

Bart274 committed:
again wrong indentation

Bart274 committed:
added test to see if function works if one device gets into the monitored zone

balloob commented:
Bart, if you would run your tests locally you will catch the indentation bugs before pushing it out

nickwaring committed:
Logic & readbility changes

amended flow when comparing devices to remove redundtant code
enhanced the way updates are made if the moniored device is no longer the closest to home
amended variable naming to consistently use "...to..." rather than "...from..."

Bart274 committed:
added tests

Bart274 committed:
remove travis build errors and updated logic

Bart274 committed:
added more tests

Bart274 committed:
remove unneeded function

Bart274 committed:
removed 3 unnecessary functions + added tests, coverage is now 72%

Bart274 committed:
updated tests, coverage 79%

Bart274 committed:
coverage at 95%

Bart274 committed:
ignored zones aren't working as they should

Bart274 committed:
logic missing when device came in ignored zone

Bart274 committed:
mistake in test

nickwaring committed:
update to proximity calculation approach

now calculates proximity for all devices and then performs comparison
minor changes to debug formatting

nickwaring committed:
fixed pylint formatting errors

nickwaring committed:
get zone friendly_name directly from the zone entity

Bart274 committed:
Revert "get zone friendly_name directly from the zone entity"

This reverts commit 43300eaf1b8e7b001bb171fd765093fe889b063c.

nickwaring committed:
error in check for all devices in zone

Bart274 committed:
fix travis build errors

Bart274 committed:
updated test for new 'stationary' direction of travel

Bart274 committed:
nearest was set to entity_id instead of friendly name

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    if 'devices' not in config[DOMAIN]:
+        _LOGGER.error('devices not found in config')
+        return False
+    else:
```
You're returning in the `if`, no need for an `else`, just put the code as usual

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+            _LOGGER.info('proximity device added: %s', variable)
+
+    # get the direction of travel tolerance from configuration.yaml
+    if 'tolerance' in config[DOMAIN]:
```
This if…else block is equal to
```python
# Define at top of file:
CONF_TOLERANCE = 'tolerance'

tolerance = config[DOMAIN].get(CONF_TOLERANCE, DEFAULT_TOLERANCE)
```

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    _LOGGER.debug('tolerance set to: %s', tolerance)
+
+    # get the zone to monitor proximity to from configuration.yaml
+    if 'zone' in config[DOMAIN]:
```
Same as above.

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+        proximity_zone = config[DOMAIN]['zone']
+    else:
+        proximity_zone = DEFAULT_PROXIMITY_ZONE
+    _LOGGER.info('zone set to %s', proximity_zone)
```
There are a lot of debug logs going on. Consolidate them at the end in 1 statement .

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+        # check for tolerance
+        if distance_travelled < tolerance * -1:
+            direction_of_travel = 'towards'
+            _LOGGER.info('%s: travelled=%s: moving=%s', entity_name,
```
Please make these calls `debug` instead of `info`

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+            return
+
+        # collect distances to the zone for all devices
+        distances_to_zone = {}
```
Soo, you could fill this dict with distances with a dict comprehension (untested):

```python
distances_to_zone = {
  state.entity_id: round(distance(…)/1000, 1) for state in
    (hass.states.get(entity_id) for entity_id in proximity_devices)
  if state is not None and state.state not in ignored_zones and ATTR_LATITUDE in state.attributes
}

closest_device = min(distances_to_zone.items(), min=lambda item: item[1])[0]
```

balloob replied:
Soo, you could fill this dict with distances with a dict comprehension (untested):

```python
distances_to_zone = {
  state.entity_id: round(distance(…)/1000, 1) for state in
    (hass.states.get(entity_id) for entity_id in proximity_devices)
  if state is not None and state.state not in ignored_zones and ATTR_LATITUDE in state.attributes
}

closest_device = min(distances_to_zone.items(), min=lambda item: item[1])[0]
```

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+        return self.dist_to
+
+    @property
+    def state_attributes(self):
```
The value that is the state should not be part of the attributes.

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    return True
+
+
+class Proximity(Entity):
```
Since your state is expressed in kilometers, you should add the property `unit_of_measurement`

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    def check_proximity_state_change(entity, old_state, new_state):
+        # pylint: disable=too-many-branches,too-many-statements,too-many-locals
+        """ Function to perform the proximity checking """
+        entity_name = new_state.attributes['friendly_name']
```
Not all entities have friendly names, use `new_state.name` instead

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+                devices_to_calculate = True
+
+            # check the location of all devices
+            if device_state.state == config[DOMAIN]['zone']:
```
This is not how you test for a zone, since zones can overlap. Use the `in_zone` method from the `zone` component instead.

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+
+            # check the location of all devices
+            if device_state.state == config[DOMAIN]['zone']:
+                device_is_in_zone = True
```
No need to keep track of this boolean. You can derive it by testing for `devices_in_zone != ''`

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+                if devices_in_zone != '':
+                    devices_in_zone = devices_in_zone + ', '
+                devices_in_zone = devices_in_zone + device_friendly
+                _LOGGER.info('%s: %s is in the monitored zone: %s',
```
Too. Much. Logging. Statements.

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    proximity.entity_id = entity_id
+
+    proximity.update_ha_state()
+    entities.add(proximity.entity_id)
```
This is not used at all, no need to maintain `entities`

balloob commented on homeassistant/components/proximity.py:
```diff
@@ -0,0 +1,325 @@
+    proximity.update_ha_state()
+    entities.add(proximity.entity_id)
+
+    def check_proximity_state_change(entity, old_state, new_state):
```
So this method should be part of the `Proximity` class. Instead of updating attributes using `proximity.dist_to` you would use `self.dist_to = …`

Bart274 committed:
some changes based on balloob's comments

Bart274 committed:
removed debug logging + added unit_of_measurement

Bart274 committed:
move method to be part of the proximity entity

Bart274 committed:
removed entities variable

nickwaring committed:
force lowercase zone names

balloob commented:
![giphy-1](https://cloud.githubusercontent.com/assets/1444314/12871264/b3fb02a0-cd26-11e5-8064-087cc18508fe.gif)

The code looks good! :dolphin: 

Final thing left to do: squash it all into 1 commit. I'll do that for you guys.

Bart274 commented:
Alright! 😊 Thanks @balloob

> Op 7 feb. 2016 om 08:12 heeft Paulus Schoutsen <notifications@github.com> het volgende geschreven:
> 
> 
> The code looks good! 
> 
> Final thing left to do: squash it all into 1 commit. I'll do that for you guys.
> 
> —
> Reply to this email directly or view it on GitHub.
>

balloob commented:
We can't rebase and squash the commits into 1 because @Bart274 merged the `dev` branch into this branch. On top of that you guys have been working in the `dev` branch on the fork. You should have branched off that branch and worked in there.

So to fix this, follow this:
![git-1](https://cloud.githubusercontent.com/assets/1444314/12871279/eae2a04c-cd27-11e5-8c97-f6f5419da9ac.png)

 - copy the files outside of your repo
 - delete your fork
 - create a new fork
 - branch off dev into a new branch (`git checkout -b proximity-component`)
 - push branch to the cloud
 - create a pull request from that new branch to HA `dev` branch.

Bart274 commented:
Ok, I will do this tomorrow :)

> Op 7 feb. 2016 om 08:20 heeft Paulus Schoutsen <notifications@github.com> het volgende geschreven:
> 
> We can't rebase and squash the commits into 1 because @Bart274 merged the dev branch into this branch. On top of that you guys have been working in the dev branch on the fork. You should have branched off that branch and worked in there.
> 
> So to fix this, follow this:
> 
> 
> copy the files outside of your repo
> delete your fork
> create a new fork
> branch off dev into a new branch (git checkout -b proximity-component)
> push branch to the cloud
> create a pull request from that new branch to HA dev branch.
> —
> Reply to this email directly or view it on GitHub.
>
